### DISCUSSION TITLE
tf.device("/cpu:0") inside the initialization.

### DISCUSSION THREAD

**Discussion by HesNobi** [keiohta/tf2rl#130] [Q&A] [answered]
Hi,
I am wondering what is the purpose of the following lines in the models initialization, and if it is nessesery?
 ```
dummy_state = tf.constant(
    np.zeros(shape=(1,)+state_shape, dtype=np.float32))
dummy_action = tf.constant(
    np.zeros(shape=(1, action_dim), dtype=np.float32))
with tf.device("/cpu:0"):
    self(dummy_state, dummy_action)
```
Thanks,

**Comment by keiohta** [comment, marked as answer]
tf2.x employs define-by-run interface that defines network shape by feeding data to the network.
Since tf2rl loads the saved weights at initialization of the Trainer class and input-output dimension can change depends on what env we want to use, defining network architecture before loading the weights enables fast check whether the saved weights match the input-output shape.

**Reply by HesNobi** [reply]
Thank you very much for the explanation. I have been using TF1 until recently. This new concept is very confusing for me. I really enjoyed the old static graph implementations, since I have a background in hardware and embedded design, and engineering. but have to adopt. :D
